sklearn/feature_extraction/image.py

```diff
@@ -125,19 +129,34 @@ def _to_graph(
     diag_idx = np.arange(n_voxels)
     i_idx = np.hstack((edges[0], edges[1]))
     j_idx = np.hstack((edges[1], edges[0]))
+
+    diag_array = diag._array if hasattr(diag, '_array') else diag
+    xp, _ = get_namespace(diag)
     graph = sparse.coo_matrix(
         (
-            np.hstack((weights, weights, diag)),
-            (np.hstack((i_idx, diag_idx)), np.hstack((j_idx, diag_idx))),
+            xp.asarray(np.hstack((weights, weights, diag_array))),
+            (xp.asarray(np.hstack((i_idx, diag_idx))), xp.asarray(np.hstack((j_idx, diag_idx)))),
```

Comment:
We an likely use [array-api stack](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html#stack-arrays-axis-0) to do something similar. (Also `numpy` has `stack`, so it should work out of the box)